### PR TITLE
Image rfo

### PR DESCRIPTION
**Pull request by AlexHeide**
The TS state code I was holding onto. Restricted Step Image Rational Function Optimization with miscellaneous fixes and tweaks to get the full Baker 1996 (not just Helgaker's subset) running smoothly.

The new algorithm isn't clearly superior to the existing in every way. I have a subset of the test results here:
[TS results.ods](https://github.com/user-attachments/files/16069626/TS.results.ods)

The new algorithm has a few less molecules that it typically struggles with compared to the `P_RFO` but neither is perfect. Most of the molecules the `RS_I_RFO` struggles with are primarly due to bad coordinate sets (STETRAZINE being the notable exception - I don't see anything noticeably wrong with those coordinates) i.e. in diels alder the two fragments are connected with stretches between C and H atoms instead of the carbon atoms where bonds are being formed. `P_RFO` tends to have more difficulty with symmetry breaking but I didn't deep dive into every `P_RFO` failure some I'm not sure what went wrong. `P_RFO` seems to often converge in 1 fewer steps than `RS_I_RFO`.

Tests currently failing because the new method isn't recognized by Psi4.

### CODE REVIEW

**Comment by psi-rking** [comment]
Excellent.  
1. I don't think the differences in iteration number are significant here.  The number of Fails is more important.
2. It gives me no delight to say that I have found performance to be basis-set and correlation dependent.  Generally speaking, these HF/tiny calculations give surfaces that are not nearly as flat (and therefore challenging) as common use situations.  I'm not saying these tests aren't reasonable to do, but I wouldn't fret too much about the details.
3. Independent of this PR, I'm vaguely concerned there is something wrong with our back-transformation or step that is causing some optimizations not to make progress, but these seem normally to involve geometrical constraints.
4. Do you know if you are using the "extra-redundant" coordinates?  I found those to be helpful, and thought they were in the code, but now glancing I don't see them.

**Comment by AlexHeide** [comment]
> 3\. Independent of this PR, I'm vaguely concerned there is something wrong with our back-transformation or step that is causing some optimizations not to make progress, but these seem normally to involve geometrical constraints.

I've definitely struggled with some issues with backtransformation issues that won't alleviate and can cause some jumping. Some of the reference iters in the test suite that I changed I believe are due to your recent PR trying to address that  - so I was hoping these issues were largely fixed. Are you still experiencing issues?


> 4\. Do you know if you are using the "extra-redundant" coordinates?  I found those to be helpful, and thought they were in the code, but now glancing I don't see them.  Update: I found it.  "add_auxiliary_bonds" which are 1-5 stretches, I think.  The default is False.  Might help.

I believe this is now reflected in `add_auxiliary_bonds()` I had not tried that in bench-marking - I'll give that a shot. That might be a sensible default to use since we are forming breaking bonds that could be on the border of considered connected. There are a number of instances where optking splits the system into multiple fragments.

With that, its maybe worth mentioning that there are a number of moelcules in the baker test suite where the atoms are not ordered contiguously in terms of fragments and optking reorders the system but this new order is not communicated back to Psi4 - that was a major headache in benchmarking.

I did also try the dimerfrag coordinates and those did not perform particularly well unfortunately.

**Comment by psi-rking** [comment]
Yes, I thought the auxiliary bonds might better describe some TS motions in particular.  My recollection is that minima optimizations performed very slightly better with them included.

You will have noted that the Bakken/Helgaker paper mentions when the back-transformation does not converge, or when the final converged geometry is farther from the target than the first one, then just go with the non-iterative first one.  I implemented this.  I don't know that it is the cause, but it is definitely indicative of an optimization in crisis.  I don't know what else to do about that except to switch to Cartesian coordinates - or perhaps to scale back the stepwise immediately (this is actually the way the old Fortran BMAT program that we used to use worked; it would always scale back until it could converge).  the internal coordinate spaces were though at that time much smaller (only a few atoms).

File: optking/exceptions.py [hunk 1]
```diff
@@ -16,9 +16,15 @@ def __init__(self, mesg="None given", err_type="Not specified"):
...
     # maybe generalize later def __init__(self, *args, **kwargs):
-    def __init__(self, mesg="None given", newLinearBends=None):
+    def __init__(self, mesg="None given", new_linear_bends=[], new_linear_torsion=[], oofp_failures=[]):
         optimize_log.error(f"AlgError: Exception created. Mesg: {mesg}")
-        if newLinearBends:
-            optimize_log.error("AlgError: New bends detected.\n")
-        self.linearBends = newLinearBends
+        if new_linear_bends:
+            optimize_log.error("AlgError: Linear bends detected.\n%s", '\n'.join(map(str, new_linear_bends)))
+        if new_linear_torsion:
```
[the remark this answers is not in the document]
Comment (AlexHeide) [reply optking/exceptions.py:23]: The tolerance for a linear bend is around 175 but we continue to allow those bends in the torsion definition until 179 which can cause two separate errors to be generated with two steps of one another we could just increase the tolerance for the torsion to match.

File: optking/displace.py [hunk 2]
```diff
@@ -425,18 +473,50 @@ def dq_to_dx(intcos, geom, dq, printDetails):
...
+        logger.debug(
+            "Cartesian step is %f times larger than internal coordinate step",
+            dx_len / dq_len
+        )
+
+        eigvals = np.linalg.eigvalsh(G)
+        largest = np.max(eigvals)
+        threshold = largest * small_val_limit
+        smallest = np.min(eigvals[eigvals > threshold])
+        new_threshold = (smallest/largest + 1e-10)
```
[the remark this answers is not in the document]
Comment (AlexHeide) [reply optking/displace.py:501]: I think this is being incorrectly flagged as new due to a merge / rebase.

File: optking/displace.py [hunk 3]
```diff
@@ -25,41 +23,64 @@
...
             # For accounting only, since displace_frag is not called.
-            dq_in[molsys.frag_intco_range(iF)] = 0
-            logger.info("\tFragment %d is frozen, so not displacing" % (iF + 1))
-        start = molsys.frag_1st_intco(iF)
-        for i, I in enumerate(F.intcos):
-            if I.frozen:
+            dq_in[molsys.frag_intco_range(i)] = 0
+            logger.info("\tFragment %d is frozen, so not displacing" % (i + 1))
+        start = molsys.frag_1st_intco(i)
+        for i, intco in enumerate(frag.intcos):
```
[the remark this answers is not in the document]
Comment (AlexHeide) [reply optking/displace.py:74]: Totally correct! I didn't catch the overlapping scopes.

**Comment by psi-rking** [comment]
@AlexHeide , what's the overall status?  Is this close to ready?

**Comment by AlexHeide** [comment]
I think its ready to go. I need to add keywords to psi4 but that should be all on that front.

**Comment by psi-rking** [comment]
Should we wait until psi4 runs successfully?  I guess the keywords can be
added first, even though they don't do anything.

**Review by psi-rking — APPROVED** [review]

**Pull request merged**